### PR TITLE
Add support for decoding Tiff images with associated alpha data

### PR DESCRIPTION
**Pull request by brianpopow**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [x] I have verified that I am following the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description

This PR adds support for decoding Tiff images with associated alpha data: The alpha data is pre-multiplied.

### CODE REVIEW

File: src/ImageSharp/Formats/Tiff/PhotometricInterpretation/Rgba16161616TiffColor{TPixel}.cs [hunk 1]
```diff
@@ -69,6 +83,12 @@ public override void Decode(ReadOnlySpan<byte> data, Buffer2D<TPixel> pixels, in
                         pixelRow,
                         pixelRow.Length);
 
+                    if (hasAssociatedAlpha)
+                    {
+                        PixelOperations<TPixel>.Instance.ToVector4(this.configuration, pixelRow, vectorsSpan);
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Tiff/PhotometricInterpretation/Rgba16161616TiffColor{TPixel}.cs:88]: Can't we use `PixelOperations<TPixel>.Instance.FromVector4Destructive` passing `PixelConversionModifiers.Premultiply` here? This allows us to use Avx2 unpremultiplication.

Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Tiff/PhotometricInterpretation/Rgba16161616TiffColor{TPixel}.cs:88]: Can't we use `PixelOperations<TPixel>.Instance.FromVector4Destructive` passing `PixelConversionModifiers.Premultiply` here? This allows us to use Avx2 unpremultiplication.

File: src/ImageSharp/Formats/Tiff/Utils/TiffUtils.cs [hunk 2]
```diff
@@ -126,6 +151,27 @@ public static TPixel ColorScaleTo32Bit<TPixel>(ulong intensity, TPixel color)
             return color;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void UnPremultiplyRow<TPixel>(Span<Vector4> vectors, Span<TPixel> pixelRow, TPixel color)
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Tiff/Utils/TiffUtils.cs:155]: This looks like we're replicating existing functionality. We have `PixelOperations` for this.

Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Tiff/Utils/TiffUtils.cs:155]: This looks like we're replicating existing functionality. We have `PixelOperations` for this.

File: src/ImageSharp/Formats/Tiff/Utils/TiffUtils.cs [hunk 3]
```diff
@@ -126,6 +151,27 @@ public static TPixel ColorScaleTo32Bit<TPixel>(ulong intensity, TPixel color)
...
+        {
+            for (int x = 0; x < vectors.Length; x++)
+            {
+                Vector4 vec = vectors[x];
+                pixelRow[x] = UnPremultiply(vec, color);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static TPixel UnPremultiply<TPixel>(Vector4 vec, TPixel color)
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Tiff/Utils/TiffUtils.cs:166]: Do you call this method when not using  collection? We have `Numerics.UnPremultiply(ref Vector4 vector)`

Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Tiff/Utils/TiffUtils.cs:166]: Do you call this method when not using  collection? We have `Numerics.UnPremultiply(ref Vector4 vector)`

**Comment by JimBobSquarePants** [comment]
@brianpopow Do you want this in 2.1? I can hold a release 'til we get it done.

**Comment by JimBobSquarePants** [comment]
@brianpopow I updated your PR to use the built in operations. I had to tweak the tolerance slightly but the output is mostly the same (better alpha blending of edge pixels with `Numerics`).

image

I'm curious as to why there is such a significant difference from the reference as I cannot see an obvious issue.

**Comment by JimBobSquarePants** [comment]
@brianpopow I might have found quite a serious issue with the Tiff decoder I hadn't spotted before. We use `FromVector4` all through the decoder codebase when we should be using `FromScaledVector4`.

**Review by JimBobSquarePants — APPROVED** [review]
LGTM 👍

**Comment by brianpopow** [comment]
> @brianpopow Do you want this in 2.1? I can hold a release 'til we get it done.

I dont think this should block 2.1 for this and it can be added for future.

> @brianpopow I might have found quite a serious issue with the Tiff decoder I hadn't spotted before. We use FromVector4 all > through the decoder codebase when we should be using FromScaledVector4. There's some cases in the Bmp decoder also.
> I'm going to update and fix in this PR.

Good spot 👍 

> Can't we use PixelOperations<TPixel>.Instance.FromVector4Destructive passing PixelConversionModifiers.Premultiply here? This allows us to use Avx2 unpremultiplication.

That was also my first thought, but I was getting different results then the reference decoder, so I was leaving it to the manual calculation.

> Can't we use PixelOperations<TPixel>.Instance.FromVector4Destructive passing PixelConversionModifiers.Premultiply here? This allows us to use Avx2 unpremultiplication.

Let me double check that, but last time I checked any difference to the reference decoder was always off by exact 1, which I blamed on floating point inaccuracies.

**Comment by JimBobSquarePants** [comment]
@brianpopow I just did a comparison using the System.Drawing reference decoder and the tests pass with the original comparison tolerance of .0.001F. This is definitely good to go.

**Pull request merged**